packages/js-client/GUIDE.md

```diff
@@ -0,0 +1,445 @@
+
+There are 3 main files that you will need to build your client:
+
+- `src/client.ts`
+- `src/context.ts`
+- `src/types.ts`
+
+### `src/types.ts`
+
+This file will include the types that your client will export, the first type that you will need will be the `ContextParams` type, this type will be used to define the parameters that your client will receive and it should be an extension of the ContextParams type from `@aragon/sdk-client-common`. This params will be later stored in the context and accessible by the modules, so if you have some endpoint, address or any other value that you need to access from your modules you should add it here.
+
+There are two additional types that you will need to add in this file.
+
+- `MyPluginState`: This type will be used to define the state of your client, this state will be stored in the context and it will be accessible by the modules. This type should extend the `ClientState` type from `@aragon/sdk-client-common`.
+
+- `MyPluginOverriddenState`: This type will be used to track what values have been modified and what values are default. This type should also extend the `OverriddenState` type from `@aragon/sdk-client-common`.
+
+In this file you should also define the state type of your client, this should extend the `ClientState` type from `@aragon/sdk-client-common`. This type will be used to define the state of your client, this state will be stored in the context and it will be accessible by the modules.
+
+```ts
+// src/types.ts
+import {ContextParams} from '@aragon/sdk-client-common';
+
+export type MyPluginClientParams = ContextParams & {
+  myCustomParam: string;
+  myCustomBackendUrl: string;
+  // Your custom params
+};
+
+export type MyPluginState = ClientState & {
+  myCustomParam: string;
+  // In the context you will be able to work with the params
+  // so the state can be a client for a service instead of the
+  // actual value
+  myCustomBackendClient: MyCustomBackendClient[];
+};
+
+export type MyPluginOverriddenState = OverriddenState & {
+  [key in keyof MyPluginContextState]: boolean;
+};
+```
+
+### `src/context.ts`
+
+The context is the responsible for storing the state of the of the client and setting the default values for the params. The context is also responsible for doing any kind of operation such as modifying some parameter, or using a parameter for creating a client for a service.
+
+This `Context` uses the previously defined `state` and `overriddenState` types.
+
+```ts
+// src/context.ts
+export class MyPluginClientContext extends Context {
+  protected state: MyPluginContextState;
+  protected overridden: MyPluginOverriddenState;
+
+  constructor(params?: Partial<MyPluginClientParams>) {
+    super();
+    this.set(contextParams);
+  }
+
+  public set(params: Partial<MyPluginClientParams>) {
+    // First we need to set the values in the parent class
+    super.set(params);
+    // Then we can set the values in our context
+    if (params.myCustomParam) {
```

Comment:
so in this set function we switch between all the things we added in `MyPluginState`

Reply:
Yep the idea is that you check that the param exists and then you set the overriden to true

Reply:
I added some comments here so it is clearer